Cython/Utility/StringTools.c

```diff
@@ -604,34 +604,64 @@ static CYTHON_INLINE PyObject* __Pyx_PyUnicode_Substring(
 #endif
 }
 
+/////////////// py_unicode_predicate.proto ///////////////
 
-/////////////// py_unicode_istitle.proto ///////////////
-
+// isprintable() is lacking C-API support in PyPy
+#if {{if method_name == 'isprintable'}}(CYTHON_COMPILING_IN_PYPY && !defined(Py_UNICODE_ISPRINTABLE)) ||{{endif}} CYTHON_COMPILING_IN_LIMITED_API
+static int __Pyx_Py_UNICODE_{{method_name.upper()}}(Py_UCS4 uchar);/*proto*/
+#else
+{{if method_name == 'istitle'}}
 // Py_UNICODE_ISTITLE() doesn't match unicode.istitle() as the latter
 // additionally allows character that comply with Py_UNICODE_ISUPPER()
-
 static CYTHON_INLINE int __Pyx_Py_UNICODE_ISTITLE(Py_UCS4 uchar) {
     return Py_UNICODE_ISTITLE(uchar) || Py_UNICODE_ISUPPER(uchar);
 }
-
-
-/////////////// py_unicode_isprintable.proto ///////////////
-
-#if CYTHON_COMPILING_IN_PYPY && !defined(Py_UNICODE_ISPRINTABLE)
-static int __Pyx_Py_UNICODE_ISPRINTABLE(Py_UCS4 uchar);/*proto*/
-#else
-#define __Pyx_Py_UNICODE_ISPRINTABLE(u)  Py_UNICODE_ISPRINTABLE(u)
+{{else}}
+#define __Pyx_Py_UNICODE_{{method_name.upper()}}(u)  Py_UNICODE_{{method_name.upper()}}(u)
+{{endif}}
 #endif
 
-/////////////// py_unicode_isprintable ///////////////
-
-#if CYTHON_COMPILING_IN_PYPY && !defined(Py_UNICODE_ISPRINTABLE)
-static int __Pyx_Py_UNICODE_ISPRINTABLE(Py_UCS4 uchar) {
+/////////////// py_unicode_predicate ///////////////
+
+{{py:
+def first_nonscaii_chr(method_name):
+    import sys
+    for i in range(128, sys.maxunicode):
+        if getattr(chr(i), method_name)():
+            return i
+    return sys.maxunicode
+
+def format_chr_for_c(i):
+    c = chr(i)
+    if c == "'":
+        return r"(Py_UCS4)'\''"
+    if c == '\\':
+        return r"(Py_UCS4)'\\'"
+    if c.isprintable():
+        return f"(Py_UCS4)'{c}'"
+    return str(i)
+}}
+
+#if {{if method_name == 'isprintable'}}(CYTHON_COMPILING_IN_PYPY && !defined(Py_UNICODE_ISPRINTABLE)) ||{{endif}} CYTHON_COMPILING_IN_LIMITED_API
+static int __Pyx_Py_UNICODE_{{method_name.upper()}}(Py_UCS4 uchar) {
     int result;
-    PyObject* py_result;
-    PyObject* ustring = PyUnicode_FromOrdinal(uchar);
+    PyObject *py_result, *ustring;
+    // Add a fast path for ascii - the switch statements get prohibitively big if we try to include
+    // all of unicode.
+    if (uchar < {{first_nonscaii_chr(method_name)}}) {
```

Comment:
Most of these are fairly small optimizations but `isdecimal` ends up as `uchar < 1632`